Add detekt-test target

davidkurkov opened the pull request:
- Add `detekt_test` rule which shares implementation with the existing `detekt-build-target`
- Modify detekt-rule implementation
    - Set default attributes that are shared between the `detekt-build-target` & `detekt-test-target`
    - Always write the `execution-result` of detekt to a text file
    - Do not fail detekt execution if `--run-as-test-target` is passed via detekt arguments
        - Running the detekt-test target will fail the execution after it evaluates the `execution-result` that is written to the output file
- Update `OneShot` execution to modify the contents of the `params-file` (if passed-in) to only passthrough detekt-related arguments

davidkurkov committed:
Add detekt test-target and test-suite

davidkurkov commented:
cc @artem-zinnatullin @bencodes

sharing this draft PR - I'm curious to get your thoughts on how I'm sharing the implementation details between the existing `detekt-build-target` and the new `detekt-test-target`...

Bencodes commented on detekt/defs.bzl:
```diff
@@ -78,65 +148,31 @@ def _impl(ctx):
+
+def detekt_test_target(name, srcs, **kwargs):
+    detekt_test(
+        name = name + "_detekt_test",
```
I would defer the naming to the caller of this macro so that we aren't making any assumptions for users of this project.

Bencodes commented on detekt/defs.bzl:
```diff
@@ -2,6 +2,69 @@
 Rule declarations.
 """
 
+_ATTRS = {
```
Little hard to tell from the Github diff, is this just a copy/paste of the original attributes or is there anything specific in here that we need to review?

davidkurkov replied:
no changes in the attributes, just a shuffle (copy-and-paste).

The only attribute change comes within the `get_default_attributes` function: https://github.com/davidkurkov/bazel_rules_detekt/pull/1/files#diff-d7c07eade8964ea97850c37c83b90f69b8b3058475560766b6d40f13aa7a8350R61-R66

Bencodes commented on detekt/defs.bzl:
```diff
@@ -78,65 +148,31 @@ def _impl(ctx):
+    attrs = _get_default_attributes(is_test_target = True),
     provides = [DefaultInfo],
     toolchains = ["@rules_detekt//detekt:toolchain_type"],
+    test = True,
```
Does this work out of the box? I believe when you mark something as `executable` or `test` you need to provide an executable for Bazel to run by writing to `ctx.outputs.executable`.

Bencodes replied:
Ah I see there is one here https://github.com/davidkurkov/bazel_rules_detekt/pull/1/files#diff-d7c07eade8964ea97850c37c83b90f69b8b3058475560766b6d40f13aa7a8350R154

davidkurkov replied:
I might be misunderstanding the question,  but there's an executable that's provided in https://github.com/davidkurkov/bazel_rules_detekt/pull/1/files#diff-d7c07eade8964ea97850c37c83b90f69b8b3058475560766b6d40f13aa7a8350R151-R156

It didn't seem to have an effect on running build with the existing `detekt-build-target` - is there something that I should pay attention to in order to verify it's not having some adverse affect on the `non-test-target`?

Bencodes commented on detekt/wrapper/src/main/java/io/buildfoundation/bazel/detekt/execute/WorkerExecutable.java:
```diff
@@ -3,6 +3,9 @@
 import io.buildfoundation.bazel.detekt.value.WorkRequest;
 import io.buildfoundation.bazel.detekt.value.WorkResponse;
 
+import java.io.*;
```
Nit: Undo this and just import the specific imports that are needed here.

davidkurkov committed:
fix imports, add comments

davidkurkov commented on detekt/wrapper/src/main/java/io/buildfoundation/bazel/detekt/Application.java:
```diff
@@ -25,6 +25,7 @@ final class OneShot implements Application {
 
         @Override
         public void run(String[] args) {
+            // TODO: handle custom argument parsing when running as one-shot
```
If the code in `WorkerExecutable` looks good, I'm thinking of extracting the pre-processing and post-processing of executing detekt out of `WorkerExecutable` in order to be able to re-use it here as part of one-shot execution.

Is there a better alternative that you'd recommend?

davidkurkov replied:
I added an `ExecutionUtils` class that is used by both `WorkerExecutable` & `OneShot`

davidkurkov committed:
use argument in implementation function instead of rule-attribute

davidkurkov committed:
Move execution helper methods to ExecutionUtils class, re-use execution utils for one-short worker execution

Bencodes commented on detekt/wrapper/src/main/java/io/buildfoundation/bazel/detekt/ExecutionUtils.java:
```diff
@@ -0,0 +1,92 @@
+        try (BufferedWriter writer = new BufferedWriter(new FileWriter(executionResultOutputPath))) {
+            writer.write(String.format("#!/bin/bash\n\nexit %d\n", exitCode));
+        } catch (IOException e) {
+            e.printStackTrace();
```
You probably don't need to write this to the console. Could just produce a bunch of noise.

Bencodes commented on detekt/wrapper/src/main/java/io/buildfoundation/bazel/detekt/ExecutionUtils.java:
```diff
@@ -0,0 +1,92 @@
+     */
+    public static void writeExecutionResultToFile(Integer exitCode, String executionResultOutputPath) {
+        try (BufferedWriter writer = new BufferedWriter(new FileWriter(executionResultOutputPath))) {
+            writer.write(String.format("#!/bin/bash\n\nexit %d\n", exitCode));
```
This should probably be written in Starlark instead of from Java. You can return the exit code written to a file and then cat that in a Starlark action that just calls exit with what ever value it reads.

davidkurkov replied:
Is this what you were thinking? https://github.com/davidkurkov/bazel_rules_detekt/pull/1/commits/1b53f5b273b2ba8f693c161e53d9271b91df30d5

davidkurkov committed:
Add special handling of overwriting params file contents for OneShot execution

davidkurkov committed:
Do not print error message to console before throwing error

davidkurkov committed:
write executable shell script with exit code using starlark

davidkurkov committed:
clean up

davidkurkov commented:
One big change since the last review: after getting OneShot execution working, I needed to make a change in OneShot to overwrite the params-file after reading/sanitizing the arguments we pass to detekt within it: https://github.com/davidkurkov/bazel_rules_detekt/pull/1/files#diff-daf3fe337b1adb03c3c9fd860bda3e8e834bd4497453eafa2caca79bdb4e07c2R30-R76

Note: OneShot execution can be forced by [setting](https://github.com/davidkurkov/bazel_rules_detekt/blob/master/detekt/defs.bzl#L75-L76) `supports-workers` & `supports-multiplex-workers` to `0`

Bencodes commented on detekt/wrapper/src/main/java/io/buildfoundation/bazel/detekt/ExecutionUtils.java:
```diff
@@ -0,0 +1,121 @@
+        String outputPath = getArgument(inputArgs, "--execution-result");
+        if (outputPath == null) {
+            System.err.println("File path for execution-result was not set");
+            System.exit(1);
```
It's generally a better idea to limit these `System#exit` calls to the main class, otherwise it becomes impossible to test classes like these.

artem-zinnatullin replied:
+1, and it's easier to follow program control flow when all System exits happen in `main`! Same with `System.err.println`, I'd say put that into `IllegalStateException` message and throw it :)

artem-zinnatullin replied:
Same with other `System.err.println()` and `System.exit`s below

Bencodes commented on detekt/wrapper/src/main/java/io/buildfoundation/bazel/detekt/execute/WorkerExecutable.java:
```diff
@@ -17,17 +21,32 @@ public Impl(Executable executable) {
         public WorkResponse execute(WorkRequest request) {
-            String[] arguments = new String[request.arguments.size()];
-            request.arguments.toArray(arguments);
+            boolean shouldRunAsTestTarget = ExecutionUtils.shouldRunAsTestTarget(request.arguments);
```
This file shouldn't need to be modified.

artem-zinnatullin commented on detekt/defs.bzl:
```diff
@@ -78,65 +141,37 @@ def _impl(ctx):
+    ctx.actions.run_shell(
+        inputs = [execution_result],
+        outputs = [final_result],
+        command = "exit_code=$(cat %s); echo -e \"#!/bin/bash\\n\\nexit $exit_code\" > %s" % (execution_result.path, final_result.path),
```
This will make `rules_detekt` incompatible with Windows, I think it's ok for now but a comment hinting that would be great, maybe later someone needing Windows support will rewrite this in cross-platform way :)

artem-zinnatullin commented on detekt/wrapper/src/main/java/io/buildfoundation/bazel/detekt/ExecutionUtils.java:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    /**
+     * Sanitizes the Detekt arguments by excluding certain arguments.
```
```suggestion
     * Sanitizes the Detekt arguments by excluding certain arguments used by the rules_detekt internally.
```

artem-zinnatullin commented on detekt/wrapper/src/main/java/io/buildfoundation/bazel/detekt/ExecutionUtils.java:
```diff
@@ -0,0 +1,121 @@
+    /**
+     * Filters out specified arguments and their corresponding values from the input argument list.
+     */
+    public static List<String> filterOutArgValuePairs(List<String> args, Set<String> excludeArgs) {
```
I'd say this needs some basic unit tests

davidkurkov commented:
Thank you for all of the input @Bencodes & @artem-zinnatullin ! I'm closing this draft PR and opened a couple of PRs against the upstream:
- https://github.com/buildfoundation/bazel_rules_detekt/pull/148
- https://github.com/buildfoundation/bazel_rules_detekt/pull/149
